### PR TITLE
Config Panel 1.0

### LINKED ISSUE #72: Config unavailable. Error: "Config panel versions '0.1' are not supported."
### Describe the bug

Choosing"Configure this app" gives an error: Config panel versions '0.1' are not supported.

### Context

- Hardware: *Old laptop or computer*
- YunoHost version: *4.3.2.2 (stable)*
- I have access to my server: *all kinds of access*. Using web UI in this case.
- Are you in a special context or did you perform some particular tweaking on your YunoHost instance?: *no*
- Using, or trying to install package version/branch: *1.0~ynh5*
 
### Steps to reproduce

 - *If the error occurs in your browser, explain what you did:*
   1. *Go to ' Applications
my_webapp '*
   2. *Click on 'Configure this app'*
   3. *See error: Config panel versions '0.1' are not supported.*

### Expected behavior

Configuration options for my_webapp should be displayed.

### Logs

Tools>Logs list does not contain this event.

### PR DESCRIPTION
**Pull request by Tagadda**
## Problem

Closes #72 

## Solution

- *And how do you fix that problem*

## PR Status

- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

**Comment by ericgaspar** [comment]
Using config panel outputs an error: https://paste.yunohost.org/raw/uweciqukon

**Comment by Tagadda** [comment]
Oh, that's because `phpversion` was none. I'll try to handle phpversion in the config panel too, that will help to hide the php-fpm config if not needed.

**Comment by Tagadda** [comment]
I'm having an issue in a config panel, where `visible` act correctly on the CLI, but not in the webadmin :(
`main.php_fpm_config.fpm_footprint` never show :/ nothing in the JS console
Something to do with `match()` I presume
https://github.com/YunoHost-Apps/my_webapp_ynh/blob/93532de252e60ecf31a8ab795e4123a8e054ab0a/config_panel.toml#L29-L34

**Comment by Tagadda** [comment]
# Manual tests

## SFTP

Add or remove `my_webapp` user from `sftp.app` group.

### Example 1 : Install without SFTP, turn SFTP on from config panel

Working great :ok:

## PHP

### Example 1 : Installation sans PHP, passage en php7.3 avec configpanel

fail : https://paste.yunohost.org/raw/bizuralani


### Example 2 : Install with php, change to another php version

- from php7.3 to :
  - php7.4 : Working great :ok: 
  - php8.0 : try to call php-fpm8.0 and fail https://paste.yunohost.org/raw/zabeboxode

- from php7.4 to :
  - php7.3 : Working great :ok: 
    - and change again php8.0 : https://paste.yunohost.org/raw/limakoyigi
  - php8.0 : KO

### Example 3 : Installation avec php7.3, passage sans PHP avec configpanel

Fail miserably because when evaluate_simple_js_expression `visible` for `main.php_fpm_config.free_footprint`, somehow `context.fpm_footprint` is `None`.
```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/moulinette/interfaces/api.py", line 460, in process
    ret = self.actionsmap.process(arguments, timeout=30, route=_route)
  File "/usr/lib/python3/dist-packages/moulinette/actionsmap.py", line 586, in process
    return func(**arguments)
  File "/usr/lib/python3/dist-packages/yunohost/log.py", line 419, in func_wrapper
    result = func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/yunohost/app.py", line 1529, in app_config_set
    return config_.set(key, value, args, args_file, operation_logger=operation_logger)
  File "/usr/lib/python3/dist-packages/yunohost/utils/config.py", line 340, in set
    self._ask()
  File "/usr/lib/python3/dist-packages/yunohost/utils/config.py", line 582, in _ask
    questions = ask_questions_and_parse_answers(
  File "/usr/lib/python3/dist-packages/yunohost/utils/config.py", line 1398, in ask_questions_and_parse_answers
    new_values = question.ask_if_needed()
  File "/usr/lib/python3/dist-packages/yunohost/utils/config.py", line 749, in ask_if_needed
    if self.visible and not evaluate_simple_js_expression(
  File "/usr/lib/python3/dist-packages/yunohost/utils/config.py", line 192, in evaluate_simple_js_expression
    return evaluate_simple_ast(node, context)
  File "/usr/lib/python3/dist-packages/yunohost/utils/config.py", line 133, in evaluate_simple_ast
    value = evaluate_simple_ast(value, context)
  File "/usr/lib/python3/dist-packages/yunohost/utils/config.py", line 146, in evaluate_simple_ast
    return re.match(
  File "/usr/lib/python3.9/re.py", line 191, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
```

**Comment by Tagadda** [comment]
The phpversion selection is broken in several scenario. Not sure how to fix it.

**Review by ericgaspar — APPROVED** [review]

**Comment by ericgaspar** [comment]
Tested with web admin. It works. (haven't checked if the PHP version is really changed thought...)
Shall we merge?

**Comment by ericgaspar** [comment]
after testing, changing PHP8.0 ti 7.4 version from config panel may have broken PHP7.4 service : https://paste.yunohost.org/sobudobola

**Comment by tituspijean** [comment]
> after testing, changing PHP8.0 ti 7.4 version from config panel may have broken PHP7.4 service : https://paste.yunohost.org/sobudobola

Can you check all the config files in `/etc/php/{7.4,8.0}/fpm/pool.d` and see if there are any duplicate sockets?

**Comment by ericgaspar** [comment]
OK, after installing with PHP7.4 and switching to PHP8.0, we end up with 2 versions of `my_webapp.conf` with PHP8.0 service error: `already listen on /var/run/php/php7.4-fpm-my_webapp.sock`
Switching from PHP7.4 to PHP8.0 should erase `/etc/php/7.4/fpm/pool.d/my_webapp.conf` (et vice-versa) but it doesn't...

File: scripts/config [hunk 1]
```diff
@@ -9,115 +9,74 @@
...
+ynh_app_config_apply() {
+    _ynh_app_config_apply
+
+
+    if [ "${changed[phpversion]}" == "true" ] && [ "$phpversion" == "none" ]
+    then
+        # FIXME: This miserably fails because the setting phpversion get his 
+        # new value before ynh_app_config_apply gets called, so 
+        # ynh_remove_fpm_config try to remove the new not-yet-existent version.
+        ynh_remove_fpm_config
```
Comment (tituspijean) [inline scripts/config:62]:
```suggestion
        new_phpversion=$phpversion
        phpversion=$(ynh_app_setting_get --app=$app --key=phpversion)
        ynh_remove_fpm_config
        phpversion=$new_phpversion
```

File: scripts/config [hunk 2]
```diff
@@ -9,115 +9,74 @@
...
+    elif [ "${changed[phpversion]}" == "true" ]
+    then
+        ynh_add_fpm_config --usage=low --footprint=low --phpversion=$phpversion
```
Comment (tituspijean) [inline scripts/config:65]: @ericgaspar I guess your issue lies at this line actually. Untested code, still lazily cheating:

```suggestion
        new_phpversion=$phpversion
        phpversion=$(ynh_app_setting_get --app=$app --key=phpversion)
        ynh_remove_fpm_config
        phpversion=$new_phpversion
        ynh_add_fpm_config --usage=low --footprint=low --phpversion=$phpversion
```

**Comment by tituspijean** [comment]
We need to use a custom setter to work around that. :)

**Comment by tituspijean** [comment]
It should be fixed! Please test. 😃

**Comment by fbeauvais** [comment]
Whats the fix for this?

**Comment by tituspijean** [comment]
> Whats the fix for this?

What do you mean?

**Comment by Tagadda** [comment]
LGTM! (Can't approve my own PR tho 😅)

**Review by tituspijean — APPROVED** [review]
Biased LGTM, and tested.

File: scripts/_common.sh [hunk 3]
```diff
@@ -365,3 +365,38 @@ ynh_app_changelog () {
...
+    local username
+    local groups
+
+    # Manage arguments with getopts
+    ynh_handle_getopts_args "$@"
+    groups="${groups:-}"
+
+	local group
+	for group in $groups; do
+		gpasswd -d "$username" "$group"
```
Comment (alexAubin) [inline scripts/_common.sh:400]: o.o

**Review by alexAubin — APPROVED** [review]
Super quick review, naive LGTM

File: scripts/_common.sh (hunk 3 above)
Comment (tituspijean) [reply scripts/_common.sh:400]: I may have shamelessly copied it from https://unix.stackexchange.com/a/29572, while there is a more Debian answer right below. :/

Comment (tituspijean) [reply scripts/_common.sh:400]: Actually... nah. It's fine like that. With `deluser` there's so many ways to f*ck everything up if `$username` or `$group` is unassigned.

(I may or may not have deleted `nobody` user from my system while testing it...)

**Pull request merged**